File: leetcode/2800/LC2841MaxSum.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LC2841MaxSum {
    public long maxSum(List<Integer> nums, int m, int k) {
        long sum  = 0, max = 0;
        int left = 0, right = k;
        Map<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < right; i++) {
            sum += nums.get(i);
            map.put(nums.get(i), map.getOrDefault(nums.get(i), 0) + 1);
        }
        if (map.size() >= m) {
            max = Math.max(sum, max);
        }

        while (right < nums.size()) {
            sum += nums.get(right);
            map.put(nums.get(right), map.getOrDefault(nums.get(right), 0) + 1);
            right++;

            sum -= nums.get(left);
            map.put(nums.get(left), map.get(nums.get(left)) - 1);
            if (map.get(nums.get(left)) == 0) {
                map.remove(nums.get(left));
            }
            left++;

            if (map.size() >= m) {
                max = Math.max(max, sum);
            }
        }

        return max;
    }
}

File: leetcode/2800/LC2841maxSum.go
func maxSum(nums []int, m int, k int) int64 {
	var sum, max int64
	left, right := 0, k
	freqMap := make(map[int]int)

	for i := 0; i < right; i++ {
		sum += int64(nums[i])
		freqMap[nums[i]]++
	}

	if len(freqMap) >= m {
		max = sum
	}

	for right < len(nums) {
		sum += int64(nums[right])
		freqMap[nums[right]]++
		right++

		sum -= int64(nums[left])
		freqMap[nums[left]]--
		if freqMap[nums[left]] == 0 {
			delete(freqMap, nums[left])
		}
		left++

		if len(freqMap) >= m && sum > max {
			max = sum
		}
	}

	return max
}